tsl/src/remote/copy_fetcher.c

```diff
@@ -240,6 +241,66 @@ copy_data_check_header(StringInfo copy_data)
 	}
 }
 
+/*
+ * End the COPY after receiving EOF or canceling a query (e.g., due to a LIMIT
+ * being reached).
+ *
+ * This should be called after canceling a query, or, after reading all data,
+ * the file trailer, and getting an EOF return value.
+ */
+static void
+end_copy(CopyFetcher *fetcher)
+{
+	PGconn *conn = remote_connection_get_pg_conn(fetcher->state.conn);
+	PGresult *final_pgres = NULL;
+	PGresult *pgres = NULL;
+	/* Expect COMMAND OK if we received EOF. If the query was canceled (e.g.,
+	 * due to a LIMIT), expect FATAL_ERROR (query failed) */
+	ExecStatusType expected_status = fetcher->state.eof ? PGRES_COMMAND_OK : PGRES_FATAL_ERROR;
+	ExecStatusType received_status;
+
+	Assert(fetcher->state.open);
+
+	/* Read results until NULL */
+	while ((pgres = PQgetResult(conn)))
+	{
+		if (final_pgres == NULL)
```

Comment:
Help me across the street here. final_pgres starts out as NULL, so the branch is true on the first time. So the first result is the final result? 

Apart of that, if we move  `received_status = PQresultStatus(final_pgres);`  into this branch and initialised `ExecStatusType received_status = PGRES_FATAL_ERROR;` , we can do away with the else branch and just call `remote_result_status(pgres)` unconditionally here.

Reply:
> Help me across the street here. final_pgres starts out as NULL, so the branch is true on the first time. So the first result is the final result?

We expect there to be only one PGResult. Maybe we should change this code to complain if there are more PGResults than one, or even simpler, just request the first result and not check what happens next. This is what `PQendCopy` does.

Reply:
Thanks for the explanation. I'll say that the intent of the code isn't too obvious from the code. Also if there are extra results, that definitely warrants a warning.

Reply:
Yes, we expect one result, but the libpq doc explicitly says: `PQgetResult must be called repeatedly until it returns a null pointer, indicating that the command is done.` So, following suggested practice here.